drivers/adc/adc_mcux_lpadc.c

```diff
@@ -0,0 +1,401 @@
+		break;
+	default:
+		LOG_ERR("Unsupported oversampling value %d",
+			sequence->oversampling);
+		return -ENOTSUP;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+	int error = adc_context_wait_for_completion(&data->ctx);
+
+	return error;
+}
+
+static int mcux_lpadc_read_async(const struct device *dev,
+			const struct adc_sequence *sequence,
+			struct k_poll_signal *async)
+{
+	struct mcux_lpadc_data *data = dev->data;
+	int error;
+
+	adc_context_lock(&data->ctx, async ? true : false, async);
+	error = mcux_lpadc_start_read(dev, sequence);
+	adc_context_release(&data->ctx, error);
+
+	return error;
+}
+
+static int mcux_lpadc_read(const struct device *dev,
+		   const struct adc_sequence *sequence)
+{
+	return mcux_lpadc_read_async(dev, sequence, NULL);
+}
+
+static void mcux_lpadc_start_channel(const struct device *dev)
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	struct mcux_lpadc_data *data = dev->data;
+
+	data->channel_id = find_lsb_set(data->channels) - 1;
+
+	LOG_DBG("Starting channel %d", data->channel_id);
+
+	lpadc_conv_command_config_t cmd_config;
+
+	LPADC_GetDefaultConvCommandConfig(&cmd_config);
+	cmd_config.channelNumber = data->channel_id;
+#if defined(FSL_FEATURE_LPADC_HAS_CMDL_MODE) \
+	&& FSL_FEATURE_LPADC_HAS_CMDL_MODE
+	cmd_config.conversionResolutionMode = data->resolution;
+#endif /* FSL_FEATURE_LPADC_HAS_CMDL_MODE */
+	cmd_config.hardwareAverageMode = data->average;
+	LPADC_SetConvCommandConfig(config->base, 1, &cmd_config);
+
+	lpadc_conv_trigger_config_t trigger_config;
+
+	LPADC_GetDefaultConvTriggerConfig(&trigger_config);
+
+	trigger_config.targetCommandId = 1;
+
+	/* configures trigger0. */
+	LPADC_SetConvTriggerConfig(config->base, 0, &trigger_config);
+
+	/* 1 is trigger0 mask. */
+	LPADC_DoSoftwareTrigger(config->base, 1);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct mcux_lpadc_data *data =
+	CONTAINER_OF(ctx, struct mcux_lpadc_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	mcux_lpadc_start_channel(data->dev);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+			  bool repeat_sampling)
+{
+	struct mcux_lpadc_data *data =
+		CONTAINER_OF(ctx, struct mcux_lpadc_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static void mcux_lpadc_isr(const struct device *dev)
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	struct mcux_lpadc_data *data = dev->data;
+	ADC_Type *base = config->base;
+
+	lpadc_conv_result_t conv_result;
+
+#if (defined(FSL_FEATURE_LPADC_FIFO_COUNT) \
+	&& (FSL_FEATURE_LPADC_FIFO_COUNT == 2U))
+	LPADC_GetConvResult(base, &conv_result, 0U);
+#else
+	LPADC_GetConvResult(base, &conv_result);
+#endif /* FSL_FEATURE_LPADC_FIFO_COUNT */
+	uint16_t result = data->ctx.sequence.resolution < 16 ?
+			conv_result.convValue >> 3 : conv_result.convValue;
```

Comment:
If I understand it correctly, the 13-bit resolution case is handled here, but what about the 12-bit one?

Reply:
The 12-bit case is also handled as the msb is always 0.
Hopefully this screenshot from the documentation shows what I mean and why the 12-bit case is handled here.
![Screenshot from 2020-10-02 15-28-03](https://user-images.githubusercontent.com/69258618/94935319-d47d1380-04c4-11eb-988d-99817e0d76ec.png)